app/controllers/twilio_webhooks_controller.rb

```diff
@@ -38,12 +38,17 @@ def create_incoming_text_message
       )
 
       if FileTypeAllowedValidator::VALID_MIME_TYPES.include? content_type
-        extension = MIME::Types[content_type].first.extensions.first
-        filename_with_extension = "#{filename}.#{extension}"
-        document.upload.attach(io: StringIO.new(Net::HTTP.get(URI(media_url))),
-                                          filename: filename_with_extension,
-                                          content_type: content_type,
-                                          identify: false)
+        response = Net::HTTP.get_response(URI(media_url)) # first we get a redirect from Twilio to S3
+        response = Net::HTTP.get_response(URI(response['location'])) # then we get a redirect from S3 to S3
+        response = Net::HTTP.get_response(URI(response['location'])) # finally we should get a 200 OK with the file
+        filename_from_s3 = response['content-disposition'].split('"').last # S3 gives us the original filename
+        document.upload.attach(
+          io: StringIO.new(response.body),
+          filename: filename_from_s3,
+          content_type: content_type,
+          identify: false
+        )
+        document.update(display_name: filename_from_s3)
```

Comment:
This is great!

One day, I would like to see the `document` auto-compute its `display_name` from the `upload` when the `upload` is created/modified. Not at all a showstopper at all.

Reply:
@alaroia-cfa it's supposed to! https://github.com/codeforamerica/vita-min/blob/master/app/models/document.rb#L46

Reply:
Then maybe we can remove line 51! Or if not, then maybe something is weird about the hook.